FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/mqtt_operations.c

```diff
@@ -0,0 +1,1069 @@
+
+                /* Cleanup the publish packet from the #pxOutgoingPublishPackets
+                 * array when a PUBACK is received. */
+                prvCleanupOutgoingPublishWithPacketID( usPacketIdentifier );
+                break;
+
+            /* Any other packet type is invalid. */
+            default:
+                LogError( ( "Unknown packet type received:(%02x).",
+                            pxPacketInfo->type ) );
+        }
+    }
+}
+/*-----------------------------------------------------------*/
+
+static bool prvHandlePublishResend( MQTTContext_t * pxMqttContext )
+{
+    bool xReturnStatus = false;
+    MQTTStatus_t xMqttStatus = MQTTSuccess;
+    uint8_t ucIndex = 0U;
+
+    configASSERT( pxOutgoingPublishPackets != NULL );
+
+    /* Resend all the QoS1 publishes still in the #pxOutgoingPublishPackets array.
+     * These are the publishes that haven't received a PUBACK yet. When a PUBACK
+     * is received, the corresponding publish is removed from the array. */
+    for( ucIndex = 0U; ucIndex < mqttopMAX_OUTGOING_PUBLISHES; ucIndex++ )
+    {
+        if( pxOutgoingPublishPackets[ ucIndex ].usPacketId != mqttopMQTT_PACKET_ID_INVALID )
+        {
+            pxOutgoingPublishPackets[ ucIndex ].xPubInfo.dup = true;
+
+            LogDebug( ( "Sending duplicate PUBLISH with packet id %u.",
+                        pxOutgoingPublishPackets[ ucIndex ].usPacketId ) );
+            xMqttStatus = MQTT_Publish( pxMqttContext,
+                                        &pxOutgoingPublishPackets[ ucIndex ].xPubInfo,
+                                        pxOutgoingPublishPackets[ ucIndex ].usPacketId );
+
+            if( xMqttStatus != MQTTSuccess )
+            {
+                LogError( ( "Sending duplicate PUBLISH for packet id %u "
+                            " failed with status %s.",
+                            pxOutgoingPublishPackets[ ucIndex ].usPacketId,
+                            MQTT_Status_strerror( xMqttStatus ) ) );
+                break;
+            }
+            else
+            {
+                LogDebug( ( "Sent duplicate PUBLISH successfully for packet id %u.",
+                            pxOutgoingPublishPackets[ ucIndex ].usPacketId ) );
+            }
+        }
+    }
+
+    /* Were all the unacknowledged QoS1 publishes successfully re-sent? */
+    if( ucIndex == mqttopMAX_OUTGOING_PUBLISHES )
+    {
+        xReturnStatus = true;
+    }
+
+    return xReturnStatus;
+}
+/*-----------------------------------------------------------*/
+
+bool xEstablishMqttSession( MQTTPublishCallback_t xPublishCallback,
+                            char * pcClientCertLabel,
+                            char * pcPrivateKeyLabel )
+{
+    bool xReturnStatus = false;
+    MQTTStatus_t xMqttStatus;
+    MQTTConnectInfo_t xConnectInfo;
+    MQTTFixedBuffer_t xNetworkBuffer;
+    TransportInterface_t xTransport;
+    bool xCreateCleanSession = false;
+    MQTTContext_t * pxMqttContext = &xMqttContext;
```

Comment:
This should be a parameter rather than refer to a global.

Reply:
These mqtt_operations changes will be addressed in a future PR, where mqtt_demo_helpers is adapted to work with PKCS11.